File: app/src/main/java/com/codebrew/clikat/module/product_detail/adapter/ProductListAdapter.java
package com.codebrew.clikat.module.product_detail.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.codebrew.clikat.R;
import com.codebrew.clikat.databinding.ItemVarientListBinding;
import com.codebrew.clikat.modal.other.VariantsBean;
import com.codebrew.clikat.utils.NpaGridLayoutManager;
import com.codebrew.clikat.utils.configurations.Configurations;

import java.util.List;

public class ProductListAdapter extends RecyclerView.Adapter<ProductListAdapter.ViewHolder> {

    private List<VariantsBean> modelList;
    private Context context;
    private ProductVarientListAdapter.FilterVarientCallback mfilterCallback;
    private String varientId;

    public ProductListAdapter(List<VariantsBean> modelList) {
        this.modelList = modelList;
    }

    public void settingCallback(ProductVarientListAdapter.FilterVarientCallback mfilterCallback) {
        this.mfilterCallback = mfilterCallback;
    }

    public void settingQuery(String adapterPosition, int adpaterPosition) {
        this.varientId = adapterPosition;

        notifyItemChanged(adpaterPosition,modelList);
        //     notifyDataSetChanged();
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        context = parent.getContext();

        ItemVarientListBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                R.layout.item_varient_list, parent, false);
        binding.setColor(Configurations.colors);
        return new ViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        int pos = holder.getAdapterPosition();

        if (modelList.get(pos).getVariant_values().size() > 0) {
            holder.tvVarientName.setText(modelList.get(pos).getVariant_name());

            ProductVarientListAdapter adapter;
            if (modelList.get(pos).getVariant_type() != null && modelList.get(pos).getVariant_type() == 1) {

                adapter = new ProductVarientListAdapter();
                adapter.settingCallback(context,mfilterCallback, "color", modelList.get(pos).getVariant_values(), pos,varientId == null);
                holder.rvVarientList.clearAnimation();
                holder.rvVarientList.setAdapter(adapter);


            } else {
                adapter = new ProductVarientListAdapter();
                adapter.settingCallback(context,mfilterCallback, "size", modelList.get(pos).getVariant_values(), pos,varientId == null);
                holder.rvVarientList.clearAnimation();
                holder.rvVarientList.setAdapter(adapter);
            }
        } else {
            holder.tvVarientName.setVisibility(View.GONE);
            holder.rvVarientList.setVisibility(View.GONE);
        }

    }

    @Override
    public int getItemCount() {
        return modelList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        RecyclerView rvVarientList;
        TextView tvVarientName;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvVarientName = itemView.findViewById(R.id.tv_varientName);
            rvVarientList = itemView.findViewById(R.id.rv_varient_list);

            LinearLayoutManager layoutManager = new NpaGridLayoutManager(context, 1, LinearLayoutManager.HORIZONTAL, false);
            rvVarientList.setLayoutManager(layoutManager);
        }
    }
}

File: ridecustomer/src/main/java/com/trava/user/ui/home/orderdetails/omcoproducts/ProductListAdapter.kt
package com.trava.user.ui.home.orderdetails.omcoproducts

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.trava.user.R
import com.trava.user.databinding.ProductlistItemlayoutBinding
import com.trava.user.utils.ConfigPOJO
import com.trava.user.webservices.models.ProductListModel
import kotlinx.android.synthetic.main.productlist_itemlayout.view.*


class ProductListAdapter (var mContext: Context?, val productModeList: ArrayList<ProductListModel>,  val  listClickCallback : ListClickCallback) : RecyclerView.Adapter<MyHolder>() {
    var state: String? = null

    interface ListClickCallback {
        fun listClickCallback(vi: View,position:Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder {
        var binding: ProductlistItemlayoutBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.context),
                R.layout.productlist_itemlayout, parent, false)
        binding.color = ConfigPOJO.Companion
        return MyHolder(binding.root)
    }

    override fun onBindViewHolder(holder: MyHolder, position: Int) {
        holder.name.setText(productModeList.get(position).product_name)
        holder.receiver.setText(productModeList.get(position).receiver_name.toString())
        holder.phonenumber.setText(productModeList.get(position).receiver_phone_number)
        holder.tvProductNumber.setText(mContext!!.getString(R.string.product)+""+(position+1))

        holder.iv_delete.setOnClickListener {
            listClickCallback.listClickCallback(it,position)
        }
     }

    override fun getItemCount(): Int {
        return productModeList.size;
    }

    public fun getList(): ArrayList<ProductListModel>
    {
        return productModeList
    }
}

open class MyHolder(view: View) : RecyclerView.ViewHolder(view) {
    var name = view.tvproductName
    var receiver = view.tvRxName
    var phonenumber = view.tvNumber
    var tvProductNumber = view.tvProductNumber
    var iv_delete = view.iv_delete
}